[FIX] mrp: correctly update MO consumption on kit BoM update

vandroogenbd opened the pull request:
#### This PR addresses 2 issues with the update of Manufacturing Order with kit products in the BoM (and subsequent bugs). Please see commits for details on the issues.

Issue 1
-----
When using the "Update BOM" button for a confirmed manufacturing order, quantities for components derived from kit products are incorrectly updated to the kit product's bom instead of the new required quantity.

Issue 2
-----
When using the "Update BOM" button for a confirmed manufacturing order, components used to produce a specific variant of a kit product are not added to the MO consumables.

Commit breakdown
-----
- Commit 1 (handle component quantities - first ticket):
  - use the correct exploded value for the quantity (divided by the original quantity as the ratio isn't done for the top-level bom)
  - update test_update_mo_from_bom_with_kit with correct values
  - extend test_update_mo_from_bom_with_kit for when the kit components' quantities are updated
- Commit 2 (handle components for specific variants - second ticket):
  - adapt the filtering of bom lines to keep them based on the kit's attribute instead of the end product one's
  - add a new test_update_mo_from_bom_with_kit_vraiants
- Commit 3 (handle confusion of multi-variants/UoM kits - additional issues):
  - change bom_lines_by_id to a defaultdict to consolidate quantity of components
  - update test_update_mo_from_bom_with_kit to test for multiple UoMs
  - update test_update_mo_from_bom_with_kit_variants to test for multiple variants

-----
Tickets:
opw-4473337
opw-4480039

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/196227.png)](https://mergebot.odoo.com/odoo/odoo/pull/196227)

lase-odoo commented on addons/mrp/tests/test_order.py:
```diff
@@ -4328,8 +4328,15 @@ def test_update_mo_from_bom_with_kit(self):
+            {'product_id': kit_bom.bom_line_ids[1].product_id.id, 'product_uom_qty': 12, 'product_uom': kit_bom.bom_line_ids[1].product_id.uom_id.id},
+        ])
+        # Check propagation upon update of the kit quantity
+        kit_bom_line.write({'product_qty': 3})
```
I would also modify the ratio of one of the bom lines to ensure it propagates correctly:
```diff
- kit_bom_line.write({'product_qty': 3})
+ kit_bom_line.product_qty = 3
+ kit_bom.bom_line_ids[0].product_qty =  4
```
Also, picky comment: When you modify a single field on a record there is no need to invoke the write method directly, you can just use the default __set__ (that will call the write of the `Fields` anyway.

vandroogenbd replied:
Done!

lase-odoo reviewed (approved):
LGTM. I was afraid that we would not trigger the change of quantity if we were to double the  product_qty on the MO + double the quantity of a bom line of the bom of the kit component since the last part of the condition `bom_line.product_qty != move_raw_qty` would fail:
https://github.com/odoo/odoo/blob/994c50671e6dc694d869fe27b385002634432dcd/addons/mrp/models/mrp_production.py#L2462-L2470
However, for kit's we are saved by the second condition: ` move_raw.bom_line_id.bom_id != bom` 

Thanks for the fix + test David ! Really, good job !

lase-odoo commented on addons/mrp/models/mrp_production.py:
```diff
@@ -2418,14 +2418,14 @@ def _link_bom(self, bom):
         ratio = self._get_ratio_between_mo_and_bom_quantities(bom)
         _dummy, bom_lines = bom.explode(self.product_id, bom.product_qty)
-        bom_lines_by_id = {(line.id, line.product_id.id): line for line, _dummy in bom_lines if filter_by_attributes(line)}
+        bom_lines_by_id = {(line.id, line.product_id.id): line for line, context in bom_lines if filter_by_attributes(line, context['product'])}
```
I would not use the terminology "context" here. This term is confusing as it makes the reader think that the info was provided via the context of the env at some point. However, since `bom_lines` is the result of the `explode` of the bom, this second argument is literally the exploded values. What about `exploded_values` instead ?

lase-odoo commented on addons/mrp/models/mrp_production.py:
```diff
@@ -2418,14 +2418,14 @@ def _link_bom(self, bom):
 
-        def filter_by_attributes(record):
-            product_attribute_ids = self.product_id.product_template_attribute_value_ids.ids
+        def filter_by_attributes(record, target_product=self.product_id):
```
`target_product` can as well be named `product` here.

Note: This should be stable since we change the spec of a method define inside the `_link_bom` and `self.product_id` is always set as the `_link_bom` starts with an ensure_one.

lase-odoo commented on addons/mrp/tests/test_order.py:
```diff
@@ -4340,6 +4340,62 @@ def test_update_mo_from_bom_with_kit(self):
         ])
 
+        # Check when adding a variant of some other kit product to the BoM
+        other_kit_product = self.env['product.template'].search([('name', '=', 'Table Kit')])
```
You should never rely on demo data's in tests.

Table Kit is a demo data it is not always present in the DB even tho it was on your set up. In particular, the test will fail on numerous runbot builds such as these without demo-data's. Create these records instead.

A general rule of thumb is that if a record is not present in the set up of your testing class and does not has a particular `ref` in `self.env` you should probably create it yourself as it is not guaranteed to exist (there are exceptions ofc).

vandroogenbd replied:
I did not know about that thanks for pointing it out!

lase-odoo reviewed (approved):
LGTM, I have left a few comments about my concerns. Thanks for your work !

vandroogenbd commented:
@ticodoo it's ready for review now! :slightly_smiling_face:

ticodoo commented on addons/mrp/tests/test_order.py:
```diff
@@ -4328,8 +4328,16 @@ def test_update_mo_from_bom_with_kit(self):
         self.assertRecordValues(mo.move_raw_ids, [
-            {'product_id': kit_bom.bom_line_ids[0].product_id.id, 'product_uom_qty': 2, 'product_uom': kit_bom.bom_line_ids[0].product_id.uom_id.id},
-            {'product_id': kit_bom.bom_line_ids[1].product_id.id, 'product_uom_qty': 3, 'product_uom': kit_bom.bom_line_ids[1].product_id.uom_id.id},
+            {'product_id': kit_bom.bom_line_ids[0].product_id.id, 'product_uom_qty': 8, 'product_uom': kit_bom.bom_line_ids[0].product_id.uom_id.id},
+            {'product_id': kit_bom.bom_line_ids[1].product_id.id, 'product_uom_qty': 12, 'product_uom': kit_bom.bom_line_ids[1].product_id.uom_id.id},
```
I don't understand why this isn't 4 of product 4 and 6 of product 3, isn't it:
2 dozens of product 5 (kit) in the bom
2 of product 4 per 1 dozen of the kit (bom_2)
3 of product 3 per 1 dozen of the kit  (bom_2)

vandroogenbd replied:
The MO quantity is 2 so it makes sense in that regard. As for why it is 2 by default, I have no idea :face_with_diagonal_mouth:

ticodoo reviewed (commented):
Seems pretty close! Just a little bit of refinement and it should be good

vandroogenbd commented:
Hi @ticodoo I made the changes you requested and modified to handle 2 things:

1. There was indeed a problem with the quantities in the MO when the main BOM was for a qty != 1. So I used it as a ratio for the exploded quantity
2. When using different variants of a kit, there was a confusion because the common components had the same bom_line ids. This was because we were using a (bom_line_id, qty) tuple as a dictionary key so the lines were overwriting the value of the dictionary entry. Since I couldn't find partial matching for Python dict keys, I changed it to a list instead.

Because of the second change however, there is a performance loss as we can't simply pop the value using a key so if you have a better approach please let me know.

Also, I added these 2 changes in a separate commit as that seemed like a cleaner approach to me.

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@vandroogenbd @ticodoo 'ci/runbot' failed on this reviewed PR.

Williambraecky commented on addons/mrp/tests/test_order.py:
```diff
@@ -4360,8 +4360,99 @@ def test_update_mo_from_bom_with_kit(self):
+            {'product_id': kit_bom.bom_line_ids[1].product_id.id, 'product_uom_qty': 45, 'product_uom': kit_bom.bom_line_ids[1].product_id.uom_id.id},
+        ])
+
+    def test_update_mo_from_bom_with_kit_variants(self):
```
Hey there @vandroogenbd @ticodoo 👋 

This test seems to be failing without demo data (most likely a missing group): <https://runbot.odoo.com/odoo/runbot.build.error/161456>
<https://runbot.odoo.com/runbot/build/77126656>

Could you take a look and provide a fix? Thanks!

vandroogenbd replied:
I'm on it, thanks for the heads up!

vandroogenbd replied:
Taken care of [here](https://github.com/odoo/odoo/pull/203225), have a great day :)

Williambraecky replied:
Thanks!
